File: 110-balanced-binary-tree/balanced-binary-tree.cpp
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return checkBalance(root) != -1;
    }

private:
    int checkBalance(TreeNode* node) {
        if (node == nullptr) {
            return 0; // Height of an empty tree is 0
        }

        int leftHeight = checkBalance(node->left);
        if (leftHeight == -1) {
            return -1; // Left subtree is unbalanced
        }

        int rightHeight = checkBalance(node->right);
        if (rightHeight == -1) {
            return -1; // Right subtree is unbalanced
        }

        int heightDiff = std::abs(leftHeight - rightHeight);
        if (heightDiff > 1) {
            return -1; // Current subtree is unbalanced
        }

        return 1 + std::max(leftHeight, rightHeight); // Return the height of the current subtree
    }
};

File: 110-balanced-binary-tree/balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        return height(root)!=-1;
    }
    private int height(TreeNode root){
        if(root == null) return 0;
        int leftHeight = height(root.left);
        if(leftHeight == -1) return -1;
        int rightHeight = height(root.right);
        if(rightHeight == -1) return -1;
        if(Math.abs(rightHeight - leftHeight) > 1)return -1;
        return Math.max(leftHeight,rightHeight) + 1;
    }
}